File: python/src/basecamp/_http.py
from __future__ import annotations

import random
import time

import httpx

from basecamp import _security
from basecamp._version import API_VERSION, VERSION
from basecamp.errors import (
    ApiError,
    AuthError,
    BasecampError,
    NetworkError,
    RateLimitError,
    UsageError,
    error_from_response,
)
from basecamp.hooks import BasecampHooks, RequestInfo, RequestResult, safe_hook


class HttpClient:
    """Sync HTTP client with retry, auth injection, and error mapping."""

    USER_AGENT = f"basecamp-sdk-python/{VERSION} (api:{API_VERSION})"

    def __init__(
        self,
        config,
        auth,
        hooks: BasecampHooks | None = None,
        *,
        user_agent: str | None = None,
        metadata: dict | None = None,
    ):
        self._config = config
        self._auth = auth
        self._hooks = hooks or BasecampHooks()
        self._metadata = metadata or {}
        self._user_agent = user_agent or self.USER_AGENT
        self._client = httpx.Client(
            timeout=httpx.Timeout(config.timeout, connect=10.0),
            follow_redirects=False,
        )

    @property
    def base_url(self) -> str:
        return self._config.base_url

    def get(self, url: str, *, params: dict | None = None) -> httpx.Response:
        url = self._build_url(url)
        return self._request_with_retry("GET", url, params=params)

    def get_absolute(self, url: str, *, params: dict | None = None) -> httpx.Response:
        if not _security.is_localhost(url):
            _security.require_https(url, "URL")
        return self._request_with_retry("GET", url, params=params)

    def post(self, url: str, *, json_body: dict | None = None, operation: str | None = None) -> httpx.Response:
        url = self._build_url(url)
        return self._mutation("POST", url, json_body=json_body, operation=operation)

    def put(self, url: str, *, json_body: dict | None = None, operation: str | None = None) -> httpx.Response:
        url = self._build_url(url)
        return self._mutation("PUT", url, json_body=json_body, operation=operation)

    def delete(self, url: str, *, operation: str | None = None) -> httpx.Response:
        url = self._build_url(url)
        return self._mutation("DELETE", url, operation=operation)

    def post_raw(
        self,
        url: str,
        *,
        content: bytes,
        content_type: str,
        params: dict | None = None,
        operation: str | None = None,
    ) -> httpx.Response:
        url = self._build_url(url)
        if operation and self._is_retryable_operation(operation):
            return self._request_with_retry(
                "POST",
                url,
                params=params,
                content=content,
                content_type=content_type,
            )
        return self._single_request(
            "POST",
            url,
            params=params,
            content=content,
            content_type=content_type,
        )

    def get_no_retry(self, url: str) -> httpx.Response:
        url = self._build_url(url)
        return self._single_request("GET", url)

    def close(self) -> None:
        self._client.close()

    # -- internal --

    def _mutation(
        self, method: str, url: str, *, json_body: dict | None = None, operation: str | None = None
    ) -> httpx.Response:
        if operation and self._is_retryable_operation(operation):
            return self._request_with_retry(method, url, json_body=json_body)
        return self._single_request(method, url, json_body=json_body)

    def _request_with_retry(
        self,
        method: str,
        url: str,
        *,
        params: dict | None = None,
        json_body: dict | None = None,
        content: bytes | None = None,
        content_type: str | None = None,
    ) -> httpx.Response:
        attempt = 0
        last_error: BasecampError | None = None

        while True:
            attempt += 1
            if attempt > self._config.max_retries + 1:
                break

            try:
                return self._single_request(
                    method,
                    url,
                    params=params,
                    json_body=json_body,
                    content=content,
                    content_type=content_type,
                    attempt=attempt,
                )
            except (RateLimitError, NetworkError, ApiError) as e:
                if not e.retryable:
                    raise
                last_error = e
                if attempt > self._config.max_retries:
                    break
                delay = self._calculate_delay(attempt, e.retry_after)
                safe_hook(
                    self._hooks.on_retry,
                    RequestInfo(method=method, url=url, attempt=attempt),
                    attempt + 1,
                    e,
                    delay,
                )
                time.sleep(delay)

        if last_error:
            raise last_error
        raise ApiError(f"Request failed after {self._config.max_retries} retries")

    def _single_request(
        self,
        method: str,
        url: str,
        *,
        params: dict | None = None,
        json_body: dict | None = None,
        content: bytes | None = None,
        content_type: str | None = None,
        attempt: int = 1,
        _retry_count: int = 0,
    ) -> httpx.Response:
        info = RequestInfo(method=method, url=url, attempt=attempt)
        safe_hook(self._hooks.on_request_start, info)
        start = time.monotonic()

        try:
            headers = self._request_headers()
            if content_type:
                headers["Content-Type"] = content_type
            self._auth.authenticate(headers)

            response = self._client.request(
                method,
                url,
                headers=headers,
                params=params,
                json=json_body,
                content=content,
            )

            if response.status_code >= 400:
                error = self._handle_error(response)
                # 401 retry with token refresh
                if isinstance(error, AuthError) and error.http_status == 401 and _retry_count < 1:
                    tp = getattr(self._auth, "token_provider", None)
                    if tp and getattr(tp, "refreshable", False) and tp.refresh():
                        return self._single_request(
                            method,
                            url,
                            params=params,
                            json_body=json_body,
                            content=content,
                            content_type=content_type,
                            attempt=attempt,
                            _retry_count=_retry_count + 1,
                        )
                raise error

            duration = time.monotonic() - start
            safe_hook(
                self._hooks.on_request_end, info, RequestResult(status_code=response.status_code, duration=duration)
            )
            return response

        except BasecampError as e:
            duration = time.monotonic() - start
            safe_hook(self._hooks.on_request_end, info, RequestResult(duration=duration, error=e))
            raise
        except httpx.HTTPError as e:
            duration = time.monotonic() - start
            error = NetworkError(f"Connection failed: {e}")
            safe_hook(self._hooks.on_request_end, info, RequestResult(duration=duration, error=error))
            raise error from e

    def _handle_error(self, response: httpx.Response) -> BasecampError:
        body = response.content[: _security.MAX_ERROR_BODY_BYTES] if response.content else None
        return error_from_response(
            response.status_code,
            body,
            dict(response.headers),
        )

    def _request_headers(self) -> dict[str, str]:
        return {
            "User-Agent": self._user_agent,
            "Accept": "application/json",
        }

    def _build_url(self, path: str) -> str:
        if path.startswith("https://"):
            return path
        if path.startswith("http://"):
            if not _security.is_localhost(path):
                raise UsageError(f"URL must use HTTPS: {path}")
            return path
        if not path.startswith("/"):
            path = f"/{path}"
        return f"{self._config.base_url}{path}"

    def _calculate_delay(self, attempt: int, server_retry_after: int | None = None) -> float:
        if server_retry_after and server_retry_after > 0:
            return float(server_retry_after)
        base = self._config.base_delay * (2 ** (attempt - 1))
        jitter = random.random() * self._config.max_jitter
        return base + jitter

    def _is_retryable_operation(self, operation: str) -> bool:
        op_meta = self._metadata.get(operation, {})
        return op_meta.get("idempotent", False)

File: ruby/lib/basecamp/http.rb
# frozen_string_literal: true

require "faraday"
require "json"
require "time"
require "uri"

module Basecamp
  # HTTP client layer with retry, backoff, and caching support.
  # This is an internal class used by Client; you typically don't use it directly.
  class Http
    # Default User-Agent header
    USER_AGENT = "basecamp-sdk-ruby/#{VERSION} (api:#{API_VERSION})".freeze

    # Normalizes Person-shaped objects in parsed JSON.
    # For objects with personable_type and a string id:
    # - Numeric strings: coerced to Integer, no system_label
    # - Non-numeric sentinels (e.g. "basecamp"): id becomes 0, system_label preserves original
    def self.normalize_person_ids(obj)
      case obj
      when Hash
        if obj.key?("personable_type") && obj["id"].is_a?(String)
          raw_id = obj["id"]
          numeric = Integer(raw_id, exception: false)
          if numeric
            obj["id"] = numeric
          else
            obj["system_label"] = raw_id
            obj["id"] = 0
          end
        end
        obj.each_value { |v| normalize_person_ids(v) }
      when Array
        obj.each { |item| normalize_person_ids(item) }
      end
    end

    # @param config [Config] configuration settings
    # @param token_provider [TokenProvider, nil] OAuth token provider (deprecated, use auth_strategy)
    # @param auth_strategy [AuthStrategy, nil] authentication strategy
    # @param hooks [Hooks] observability hooks
    def initialize(config:, token_provider: nil, auth_strategy: nil, hooks: nil)
      @config = config
      @auth_strategy = auth_strategy || BearerAuth.new(token_provider)
      @token_provider = token_provider || (@auth_strategy.is_a?(BearerAuth) ? @auth_strategy.token_provider : nil)
      @hooks = hooks || NoopHooks.new
      @faraday = build_faraday_client
    end

    # @return [String] the configured base URL
    def base_url
      @config.base_url
    end

    # Performs a GET request.
    # @param path [String] URL path
    # @param params [Hash] query parameters
    # @return [Response]
    def get(path, params: {})
      request(:get, path, params: params)
    end

    # Performs a GET request to an absolute URL.
    # Used for endpoints not on the base API (e.g., Launchpad).
    # @param url [String] absolute URL
    # @param params [Hash] query parameters
    # @return [Response]
    def get_absolute(url, params: {})
      request(:get, url, params: params)
    end

    # Performs a POST request.
    # @param path [String] URL path
    # @param body [Hash, nil] request body
    # @return [Response]
    def post(path, body: nil)
      request(:post, path, body: body)
    end

    # Performs a PUT request.
    # @param path [String] URL path
    # @param body [Hash, nil] request body
    # @return [Response]
    def put(path, body: nil)
      request(:put, path, body: body)
    end

    # Performs a DELETE request.
    # @param path [String] URL path
    # @return [Response]
    def delete(path)
      request(:delete, path)
    end

    # Performs a POST request with raw binary data.
    # Used for file uploads (attachments).
    # @param path [String] URL path
    # @param body [String, IO] raw binary data
    # @param content_type [String] MIME content type
    # @return [Response]
    def post_raw(path, body:, content_type:)
      url = build_url(path)
      single_request_raw(:post, url, body: body, content_type: content_type, attempt: 1)
    end

    # Performs a PUT request with raw binary data.
    # Used for multipart uploads (e.g., account logo).
    # @param path [String] URL path
    # @param body [String, IO] raw binary data
    # @param content_type [String] MIME content type
    # @return [Response]
    def put_raw(path, body:, content_type:)
      url = build_url(path)
      single_request_raw(:put, url, body: body, content_type: content_type, attempt: 1)
    end

    # Performs a GET request without retry logic.
    # Used for the download flow where retry is not appropriate.
    # @param url [String] absolute URL
    # @return [Response]
    def get_no_retry(url)
      single_request(:get, url, params: {}, body: nil, attempt: 1)
    end

    # Fetches all pages of a paginated resource.
    # @param path [String] initial URL path
    # @param params [Hash] query parameters
    # @yield [Hash] each item from the response
    # @return [Enumerator] if no block given
    def paginate(path, params: {}, &block)
      return to_enum(:paginate, path, params: params) unless block

      base_url = build_url(path)
      url = base_url
      page = 0

      loop do
        page += 1
        break if page > @config.max_pages

        @hooks.on_paginate(url, page)
        response = get(url, params: page == 1 ? params : {})

        Security.check_body_size!(response.body, Security::MAX_RESPONSE_BODY_BYTES)

        begin
          items = JSON.parse(response.body)
          Http.normalize_person_ids(items)
        rescue JSON::ParserError => e
          raise Basecamp::ApiError.new("Failed to parse paginated response (page #{page}): #{Security.truncate(e.message)}")
        end
        items.each(&block)

        next_url = parse_next_link(response.headers["Link"])
        break if next_url.nil?

        next_url = Security.resolve_url(url, next_url)

        unless Security.same_origin?(next_url, base_url)
          raise Basecamp::ApiError.new(
            "Pagination Link header points to different origin: #{Security.truncate(next_url)}"
          )
        end

        url = next_url
      end
    end

    # Fetches all pages of a paginated resource, extracting items from a key.
    # Use this for endpoints that return objects like { "events": [...] }.
    # @param path [String] initial URL path
    # @param key [String] the key containing the array of items
    # @param params [Hash] query parameters
    # @yield [Hash] each item from the response
    # @return [Enumerator] if no block given
    def paginate_key(path, key:, params: {}, &block)
      return to_enum(:paginate_key, path, key: key, params: params) unless block

      base_url = build_url(path)
      url = base_url
      page = 0

      loop do
        page += 1
        break if page > @config.max_pages

        @hooks.on_paginate(url, page)
        response = get(url, params: page == 1 ? params : {})

        Security.check_body_size!(response.body, Security::MAX_RESPONSE_BODY_BYTES)

        begin
          data = JSON.parse(response.body)
          Http.normalize_person_ids(data)
        rescue JSON::ParserError => e
          raise Basecamp::ApiError.new("Failed to parse paginated response (page #{page}): #{Security.truncate(e.message)}")
        end
        unless data.key?(key)
          warn "[Basecamp SDK] paginate_key: expected key '#{key}' not found in response (page #{page})"
        end
        items = data[key] || []
        items.each(&block)

        next_url = parse_next_link(response.headers["Link"])
        break if next_url.nil?

        next_url = Security.resolve_url(url, next_url)

        unless Security.same_origin?(next_url, base_url)
          raise Basecamp::ApiError.new(
            "Pagination Link header points to different origin: #{Security.truncate(next_url)}"
          )
        end

        url = next_url
      end
    end

    # Fetches a wrapped paginated resource, returning wrapper fields + lazy paginated items.
    # Use this for endpoints that return {wrapper_field: ..., key: [items]} on every page.
    # @param path [String] initial URL path
    # @param key [String] the key containing the array of paginated items
    # @param params [Hash] query parameters
    # @return [Hash] wrapper fields merged with key => Enumerator of all items
    def paginate_wrapped(path, key:, params: {})
      base_url = build_url(path)

      @hooks.on_paginate(base_url, 1)
      first_response = get(base_url, params: params)
      Security.check_body_size!(first_response.body, Security::MAX_RESPONSE_BODY_BYTES)

      begin
        first_data = JSON.parse(first_response.body)
        Http.normalize_person_ids(first_data)
      rescue JSON::ParserError => e
        raise Basecamp::ApiError.new(
          "Failed to parse paginated response (page 1): #{Security.truncate(e.message)}"
        )
      end

      wrapper = first_data.reject { |k, _| k == key }
      first_items = first_data[key] || []

      events = Enumerator.new do |yielder|
        first_items.each { |item| yielder << item }

        next_link = parse_next_link(first_response.headers["Link"])
        url = base_url
        page = 1

        while next_link && page < @config.max_pages
          page += 1
          next_url = Security.resolve_url(url, next_link)

          unless Security.same_origin?(next_url, base_url)
            raise Basecamp::ApiError.new(
              "Pagination Link header points to different origin: " \
              "#{Security.truncate(next_url)}"
            )
          end

          @hooks.on_paginate(next_url, page)
          response = get(next_url)
          Security.check_body_size!(response.body, Security::MAX_RESPONSE_BODY_BYTES)

          begin
            data = JSON.parse(response.body)
            Http.normalize_person_ids(data)
          rescue JSON::ParserError => e
            raise Basecamp::ApiError.new(
              "Failed to parse paginated response (page #{page}): " \
              "#{Security.truncate(e.message)}"
            )
          end

          items = data[key] || []
          items.each { |item| yielder << item }

          next_link = parse_next_link(response.headers["Link"])
          url = next_url
        end
      end

      wrapper.merge(key => events)
    end

    private

    def build_faraday_client
      Faraday.new(url: @config.base_url) do |f|
        f.options.timeout = @config.timeout
        f.options.open_timeout = 10
        f.request :json
        f.response :raise_error
        f.adapter Faraday.default_adapter
      end
    end

    def request(method, path, params: {}, body: nil)
      url = build_url(path)

      # Mutations don't retry on 429/5xx to avoid duplicating data
      if method == :get
        request_with_retry(method, url, params: params)
      else
        single_request(method, url, params: params, body: body, attempt: 1)
      end
    end

    def request_with_retry(method, url, params: {})
      attempt = 0
      last_error = nil

      loop do
        attempt += 1
        break if attempt > @config.max_retries

        begin
          return single_request(method, url, params: params, body: nil, attempt: attempt)
        rescue Basecamp::RateLimitError, Basecamp::NetworkError, Basecamp::ApiError => e
          raise e unless e.retryable?

          last_error = e

          # Don't sleep if this was the last attempt
          break if attempt >= @config.max_retries

          delay = calculate_delay(attempt, e.retry_after)

          @hooks.on_retry(RequestInfo.new(method: method.to_s.upcase, url: url, attempt: attempt), attempt + 1, e,
                          delay)
          sleep(delay)
        end
      end

      raise last_error || Basecamp::ApiError.new("Request failed after #{@config.max_retries} retries")
    end

    def single_request(method, url, params:, body:, attempt:, retry_count: 0)
      info = RequestInfo.new(method: method.to_s.upcase, url: url, attempt: attempt)
      @hooks.on_request_start(info)

      start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

      begin
        response = @faraday.run_request(method, url, body, request_headers) do |req|
          req.params.merge!(params) if params.any?
        end

        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        result = RequestResult.new(status_code: response.status, duration: duration)
        @hooks.on_request_end(info, result)

        Response.new(
          body: response.body,
          status: response.status,
          headers: response.headers
        )
      rescue Faraday::ServerError, Faraday::ClientError => e
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        error = handle_error(e)
        result = RequestResult.new(
          status_code: e.response&.dig(:status),
          duration: duration,
          error: error,
          retry_after: error.respond_to?(:retry_after) ? error.retry_after : nil
        )
        @hooks.on_request_end(info, result)

        # After a successful token refresh on 401, retry the request once
        if error.is_a?(Basecamp::AuthError) && error.http_status == 401 && retry_count < 1 && @token_refreshed
          @token_refreshed = false
          return single_request(method, url, params: params, body: body, attempt: attempt, retry_count: retry_count + 1)
        end

        raise error
      rescue Faraday::Error => e
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        error = Basecamp::NetworkError.new("Connection failed", cause: e)
        result = RequestResult.new(duration: duration, error: error)
        @hooks.on_request_end(info, result)
        raise error
      end
    end

    def request_headers
      headers = {
        "User-Agent" => USER_AGENT,
        "Accept" => "application/json"
      }
      @auth_strategy.authenticate(headers)
      headers
    end

    def single_request_raw(method, url, body:, content_type:, attempt:)
      info = RequestInfo.new(method: method.to_s.upcase, url: url, attempt: attempt)
      @hooks.on_request_start(info)

      start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)

      begin
        headers = request_headers.merge("Content-Type" => content_type)
        response = @faraday.run_request(method, url, body, headers)

        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        result = RequestResult.new(status_code: response.status, duration: duration)
        @hooks.on_request_end(info, result)

        Response.new(
          body: response.body,
          status: response.status,
          headers: response.headers
        )
      rescue Faraday::ServerError, Faraday::ClientError => e
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        error = handle_error(e)
        result = RequestResult.new(
          status_code: e.response&.dig(:status),
          duration: duration,
          error: error,
          retry_after: error.respond_to?(:retry_after) ? error.retry_after : nil
        )
        @hooks.on_request_end(info, result)
        raise error
      rescue Faraday::Error => e
        duration = Process.clock_gettime(Process::CLOCK_MONOTONIC) - start_time
        error = Basecamp::NetworkError.new("Connection failed", cause: e)
        result = RequestResult.new(duration: duration, error: error)
        @hooks.on_request_end(info, result)
        raise error
      end
    end

    def handle_error(error)
      status = error.response&.dig(:status)
      body = error.response&.dig(:body)
      headers = error.response&.dig(:headers) || {}

      retry_after = parse_retry_after(headers["Retry-After"] || headers["retry-after"])
      request_id = headers["X-Request-Id"] || headers["x-request-id"]

      err = case status
      when 401
        # Try token refresh; flag for caller to retry
        @token_refreshed = @token_provider&.refreshable? && @token_provider.refresh
        Basecamp::AuthError.new("Authentication failed")
      when 403
        Basecamp::ForbiddenError.new("Access denied")
      when 404
        reason = headers["Reason"] || headers["reason"]
        if reason == "API Disabled"
          Basecamp::ApiDisabledError.new
        elsif reason == "Account Inactive"
          Basecamp::NotFoundError.new(message: "Account is inactive", hint: "The account may have an expired trial or be suspended")
        else
          message = Security.truncate(Basecamp.parse_error_message(body) || "Not found")
          Basecamp::NotFoundError.new(message: message)
        end
      when 429
        Basecamp::RateLimitError.new(retry_after: retry_after)
      when 400, 422
        message = Security.truncate(Basecamp.parse_error_message(body) || "Validation failed")
        Basecamp::ValidationError.new(message, http_status: status)
      when 500
        Basecamp::ApiError.new("Server error (500)", http_status: 500, retryable: true)
      when 502, 503, 504
        Basecamp::ApiError.new("Gateway error (#{status})", http_status: status, retryable: true)
      else
        message = Security.truncate(Basecamp.parse_error_message(body) || "Request failed (HTTP #{status})")
        Basecamp::ApiError.from_status(status || 0, message)
      end

      err.instance_variable_set(:@request_id, request_id) if request_id
      err
    end

    def build_url(path)
      if path.start_with?("https://")
        return path
      elsif path.start_with?("http://")
        raise Basecamp::UsageError.new("URL must use HTTPS: #{path}")
      end

      path = "/#{path}" unless path.start_with?("/")
      "#{@config.base_url}#{path}"
    end

    def calculate_delay(attempt, server_retry_after)
      return server_retry_after if server_retry_after&.positive?

      # Exponential backoff: base_delay * 2^(attempt-1) + jitter
      base = @config.base_delay * (2**(attempt - 1))
      jitter = rand * @config.max_jitter
      base + jitter
    end

    def parse_retry_after(value)
      return nil if value.nil? || value.empty?

      # Try parsing as seconds (integer)
      seconds = Integer(value, exception: false)
      return seconds if seconds&.positive?

      # Try parsing as HTTP-date
      begin
        date = Time.httpdate(value)
        diff = (date - Time.now).to_i
        return diff if diff.positive?
      rescue ArgumentError
        # Not a valid HTTP-date
      end

      nil
    end

    def parse_next_link(link_header)
      return nil if link_header.nil? || link_header.empty?

      link_header.split(",").each do |part|
        part = part.strip
        next unless part.include?('rel="next"')

        match = part.match(/<([^>]+)>/)
        return match[1] if match
      end

      nil
    end
  end

  # Wraps an HTTP response.
  class Response
    # @return [String] response body
    attr_reader :body

    # @return [Integer] HTTP status code
    attr_reader :status

    # @return [Hash] response headers
    attr_reader :headers

    def initialize(body:, status:, headers:)
      @body = body
      @status = status
      @headers = headers
    end

    # Parses the response body as JSON, normalizing Person-shaped objects.
    # @return [Hash, Array]
    def json
      @json ||= begin
        Security.check_body_size!(@body, Security::MAX_RESPONSE_BODY_BYTES)
        result = JSON.parse(@body)
        Http.normalize_person_ids(result)
        result
      end
    end

    # Returns whether the response was successful (2xx).
    # @return [Boolean]
    def success?
      status >= 200 && status < 300
    end
  end
end
